File: python/xxe/product_import.py
import xml.etree.ElementTree as ET
from lxml import etree
from flask import Flask, request

app = Flask(__name__)


# VULN 1: XXE via lxml parser - external entity expansion enabled by default
@app.route("/products/import-xml", methods=["POST"])
def import_products_xml():
    xml_data = request.get_data()
    parser = etree.XMLParser(resolve_entities=True)
    tree = etree.fromstring(xml_data, parser)
    products = []
    for product in tree.findall("product"):
        products.append({
            "name": product.findtext("name"),
            "price": product.findtext("price"),
        })
    return {"imported": len(products)}


# VULN 2: XXE in order processing - XML from external source parsed unsafely
@app.route("/orders/import-xml", methods=["POST"])
def import_orders_xml():
    xml_content = request.files["orders_file"].read()
    root = ET.fromstring(xml_content)
    orders = [{"id": o.findtext("id"), "total": o.findtext("total")} for o in root]
    return {"orders": orders}


# VULN 3: XXE in supplier catalog sync - unsafe XML parsing with entity resolution
def sync_supplier_catalog(xml_string):
    parser = etree.XMLParser(load_dtd=True, no_network=False, resolve_entities=True)
    root = etree.fromstring(xml_string.encode(), parser)
    return [el.text for el in root.findall(".//product/name")]

File: javascript/xxe/productImport.js
const express = require('express');
const libxmljs = require('libxmljs');
const xml2js = require('xml2js');
const { XMLParser } = require('fast-xml-parser');

const router = express.Router();

// VULN 1: libxmljs.parseXmlString() with external entity processing enabled - XXE
router.post('/products/import/xml', (req, res) => {
    const { xmlData } = req.body;
    // noent: false (default) and dtdload: true allow external entity expansion
    const xmlDoc = libxmljs.parseXmlString(xmlData, {
        noent: false,
        dtdload: true,
        dtdvalid: true,
    });
    const products = xmlDoc.find('//product').map((node) => ({
        sku: node.get('sku') ? node.get('sku').text() : '',
        name: node.get('name') ? node.get('name').text() : '',
        price: node.get('price') ? parseFloat(node.get('price').text()) : 0,
    }));
    res.json({ imported: products.length, products });
});

// VULN 2: xml2js.parseString() with no entity protection - XXE via entity expansion
router.post('/catalog/import', (req, res) => {
    const { xmlData } = req.body;
    // xml2js with explicitArray does not strip external entities from DTD declarations
    xml2js.parseString(xmlData, { explicitArray: false, explicitRoot: false }, (err, result) => {
        if (err) return res.status(400).json({ error: err.message });
        const items = result.catalog ? result.catalog.item : [];
        const catalogItems = Array.isArray(items) ? items : [items];
        res.json({ imported: catalogItems.length, catalog: catalogItems });
    });
});

// VULN 3: fast-xml-parser with processEntities: true and external entity data - XXE
router.post('/inventory/sync', (req, res) => {
    const { xmlData } = req.body;
    const parser = new XMLParser({
        ignoreAttributes: false,
        processEntities: true,
        allowBooleanAttributes: true,
    });
    // processEntities: true allows external entity references to read local files
    const inventoryData = parser.parse(xmlData);
    const items = inventoryData.inventory ? inventoryData.inventory.item : [];
    const inventory = Array.isArray(items) ? items : [items];
    res.json({ synced: inventory.length, inventory });
});

module.exports = router;
